### ISSUE TITLE
ERROR when build the ginkgo example with MFEM and CUDA

### ISSUE THREAD

**Issue by nghiakvnvsd** [ginkgo-project/ginkgo#616]
Hello everyone!

I have a problem when I try doing run the ginkgo example with MFEM. Firstly, I set the option MFEM_USE_GINKGO=YES and I run the example so it's ok but when I set both the options MFEM_USE_GINKGO=YES and MFEM_USE_CUDA=YES and run the ginkgo example. It reports:

```
nvcc  -O3 -std=c++11 -x=cu --expt-extended-lambda -arch=sm_60 -ccbin g++ -I../.. -isystem ../../../ginkgo/install/include ex1.cpp -o ex1 -L../.. -lmfem -Xlinker=-rpath,../../../ginkgo/install/lib -L../../../ginkgo/install/lib -lginkgo -lginkgo_omp -lginkgo_cuda -lginkgo_reference  -lrt
/usr/local/include/ginkgo/core/base/range.hpp(385): warning: __host__ annotation is ignored on a function("range") that is explicitly defaulted on its first declaration

/usr/local/include/ginkgo/core/base/range.hpp(385): warning: __device__ annotation is ignored on a function("range") that is explicitly defaulted on its first declaration

/usr/local/include/ginkgo/core/base/range.hpp(385): warning: __host__ annotation is ignored on a function("range") that is explicitly defaulted on its first declaration

/usr/local/include/ginkgo/core/base/range.hpp(385): warning: __device__ annotation is ignored on a function("range") that is explicitly defaulted on its first declaration

../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::GinkgoIterativeSolverBase::GinkgoIterativeSolverBase(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x57f): undefined reference to `gko::stop::ResidualNormReduction<double>::parameters_type const& gko::stop::ResidualNormReduction<double>::parameters_type::with_reduction_factor<double&>(double&) const'
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x6e7): undefined reference to `gko::stop::Iteration::parameters_type const& gko::stop::Iteration::parameters_type::with_max_iters<int&>(int&) const'
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x76a): undefined reference to `gko::stop::Combined::parameters_type const& gko::stop::Combined::parameters_type::with_criteria<std::shared_ptr<gko::stop::ResidualNormReduction<double>::Factory>&, std::unique_ptr<gko::stop::Iteration::Factory, std::default_delete<gko::stop::Iteration::Factory> > >(std::shared_ptr<gko::stop::ResidualNormReduction<double>::Factory>&, std::unique_ptr<gko::stop::Iteration::Factory, std::default_delete<gko::stop::Iteration::Factory> >&&) const'
../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::CGSolver::CGSolver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double, gko::LinOpFactory const*)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0xfd4): undefined reference to `gko::solver::Cg<double>::parameters_type const& gko::solver::Cg<double>::parameters_type::with_criteria<std::shared_ptr<gko::stop::Combined::Factory>&>(std::shared_ptr<gko::stop::Combined::Factory>&) const'
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0xfe1): undefined reference to `gko::solver::Cg<double>::parameters_type const& gko::solver::Cg<double>::parameters_type::with_preconditioner<gko::LinOpFactory const*&>(gko::LinOpFactory const*&) const'
../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::CGSolver::CGSolver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x117f): undefined reference to `gko::solver::Cg<double>::parameters_type const& gko::solver::Cg<double>::parameters_type::with_criteria<std::shared_ptr<gko::stop::Combined::Factory>&>(std::shared_ptr<gko::stop::Combined::Factory>&) const'
../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::BICGSTABSolver::BICGSTABSolver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double, gko::LinOpFactory const*)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x1324): undefined reference to `gko::solver::Bicgstab<double>::parameters_type const& gko::solver::Bicgstab<double>::parameters_type::with_criteria<std::shared_ptr<gko::stop::Combined::Factory>&>(std::shared_ptr<gko::stop::Combined::Factory>&) const'
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x1331): undefined reference to `gko::solver::Bicgstab<double>::parameters_type const& gko::solver::Bicgstab<double>::parameters_type::with_preconditioner<gko::LinOpFactory const*&>(gko::LinOpFactory const*&) const'
../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::BICGSTABSolver::BICGSTABSolver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x14cf): undefined reference to `gko::solver::Bicgstab<double>::parameters_type const& gko::solver::Bicgstab<double>::parameters_type::with_criteria<std::shared_ptr<gko::stop::Combined::Factory>&>(std::shared_ptr<gko::stop::Combined::Factory>&) const'
../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::CGSSolver::CGSSolver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double, gko::LinOpFactory const*)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x1674): undefined reference to `gko::solver::Cgs<double>::parameters_type const& gko::solver::Cgs<double>::parameters_type::with_criteria<std::shared_ptr<gko::stop::Combined::Factory>&>(std::shared_ptr<gko::stop::Combined::Factory>&) const'
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x1681): undefined reference to `gko::solver::Cgs<double>::parameters_type const& gko::solver::Cgs<double>::parameters_type::with_preconditioner<gko::LinOpFactory const*&>(gko::LinOpFactory const*&) const'
../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::CGSSolver::CGSSolver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x181f): undefined reference to `gko::solver::Cgs<double>::parameters_type const& gko::solver::Cgs<double>::parameters_type::with_criteria<std::shared_ptr<gko::stop::Combined::Factory>&>(std::shared_ptr<gko::stop::Combined::Factory>&) const'
../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::FCGSolver::FCGSolver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double, gko::LinOpFactory const*)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x19c4): undefined reference to `gko::solver::Fcg<double>::parameters_type const& gko::solver::Fcg<double>::parameters_type::with_criteria<std::shared_ptr<gko::stop::Combined::Factory>&>(std::shared_ptr<gko::stop::Combined::Factory>&) const'
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x19d1): undefined reference to `gko::solver::Fcg<double>::parameters_type const& gko::solver::Fcg<double>::parameters_type::with_preconditioner<gko::LinOpFactory const*&>(gko::LinOpFactory const*&) const'
../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::FCGSolver::FCGSolver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x1b6f): undefined reference to `gko::solver::Fcg<double>::parameters_type const& gko::solver::Fcg<double>::parameters_type::with_criteria<std::shared_ptr<gko::stop::Combined::Factory>&>(std::shared_ptr<gko::stop::Combined::Factory>&) const'
../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::GMRESSolver::GMRESSolver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double, gko::LinOpFactory const*)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x1d19): undefined reference to `gko::solver::Gmres<double>::parameters_type const& gko::solver::Gmres<double>::parameters_type::with_krylov_dim<int&>(int&) const'
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x1d25): undefined reference to `gko::solver::Gmres<double>::parameters_type const& gko::solver::Gmres<double>::parameters_type::with_criteria<std::shared_ptr<gko::stop::Combined::Factory>&>(std::shared_ptr<gko::stop::Combined::Factory>&) const'
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x1d32): undefined reference to `gko::solver::Gmres<double>::parameters_type const& gko::solver::Gmres<double>::parameters_type::with_preconditioner<gko::LinOpFactory const*&>(gko::LinOpFactory const*&) const'
../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::GMRESSolver::GMRESSolver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x1ede): undefined reference to `gko::solver::Gmres<double>::parameters_type const& gko::solver::Gmres<double>::parameters_type::with_krylov_dim<int&>(int&) const'
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x1eea): undefined reference to `gko::solver::Gmres<double>::parameters_type const& gko::solver::Gmres<double>::parameters_type::with_criteria<std::shared_ptr<gko::stop::Combined::Factory>&>(std::shared_ptr<gko::stop::Combined::Factory>&) const'
../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::IRSolver::IRSolver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double, gko::LinOpFactory const*)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x2094): undefined reference to `gko::solver::Ir<double>::parameters_type const& gko::solver::Ir<double>::parameters_type::with_criteria<std::shared_ptr<gko::stop::Combined::Factory>&>(std::shared_ptr<gko::stop::Combined::Factory>&) const'
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x20a1): undefined reference to `gko::solver::Ir<double>::parameters_type const& gko::solver::Ir<double>::parameters_type::with_solver<gko::LinOpFactory const*&>(gko::LinOpFactory const*&) const'
../../libmfem.a(ginkgo.o): In function `mfem::GinkgoWrappers::IRSolver::IRSolver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, double, double)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text+0x223f): undefined reference to `gko::solver::Ir<double>::parameters_type const& gko::solver::Ir<double>::parameters_type::with_criteria<std::shared_ptr<gko::stop::Combined::Factory>&>(std::shared_ptr<gko::stop::Combined::Factory>&) const'
../../libmfem.a(ginkgo.o): In function `std::shared_ptr<gko::AbstractFactory<gko::stop::Criterion, gko::stop::CriterionArgs> const> gko::stop::combine<std::vector<std::shared_ptr<gko::AbstractFactory<gko::stop::Criterion, gko::stop::CriterionArgs> const>, std::allocator<std::shared_ptr<gko::AbstractFactory<gko::stop::Criterion, gko::stop::CriterionArgs> const> > > >(std::vector<std::shared_ptr<gko::AbstractFactory<gko::stop::Criterion, gko::stop::CriterionArgs> const>, std::allocator<std::shared_ptr<gko::AbstractFactory<gko::stop::Criterion, gko::stop::CriterionArgs> const> > >&&)':
tmpxft_000032c2_00000000-5_ginkgo.cudafe1.cpp:(.text._ZN3gko4stop7combineISt6vectorISt10shared_ptrIKNS_15AbstractFactoryINS0_9CriterionENS0_13CriterionArgsEEEESaIS9_EEEES9_OT_[_ZN3gko4stop7combineISt6vectorISt10shared_ptrIKNS_15AbstractFactoryINS0_9CriterionENS0_13CriterionArgsEEEESaIS9_EEEES9_OT_]+0x83): undefined reference to `gko::stop::Combined::parameters_type const& gko::stop::Combined::parameters_type::with_criteria<std::vector<std::shared_ptr<gko::AbstractFactory<gko::stop::Criterion, gko::stop::CriterionArgs> const>, std::allocator<std::shared_ptr<gko::AbstractFactory<gko::stop::Criterion, gko::stop::CriterionArgs> const> > > >(std::vector<std::shared_ptr<gko::AbstractFactory<gko::stop::Criterion, gko::stop::CriterionArgs> const>, std::allocator<std::shared_ptr<gko::AbstractFactory<gko::stop::Criterion, gko::stop::CriterionArgs> const> > >&&) const'
collect2: error: ld returned 1 exit status
makefile:42: recipe for target 'ex1' failed
```

I have no idea about it. May you help me to fix it?

Thank you!

**Comment by tcojean** [comment]
One quick test, instead of launching `make` directly, can you try to compile the example by hand? I think it is related to #586 which isn't fixed/merged yet.
```bash
g++ -O3 -std=c++11 -I../.. -isystem ../../../ginkgo/install/include ex1.cpp -o ex1 -L../.. -lmfem -Wl,-rpath,../../../ginkgo/install/lib -L../../../ginkgo/install/lib -lginkgo -lginkgo_omp -lginkgo_cuda -lginkgo_reference -lrt
```

**Comment by nghiakvnvsd** [comment]
> One quick test, instead of launching `make` directly, can you try to compile the example by hand? I think it is related to #586 which isn't fixed/merged yet.
> Once you are in the correct directory (where there is `ex1.cpp`):
> 
> ```shell
> g++ -O3 -std=c++11 -I../.. -isystem ../../../ginkgo/install/include ex1.cpp -o ex1 -L../.. -lmfem -Wl,-rpath,../../../ginkgo/install/lib -L../../../ginkgo/install/lib -lginkgo -lginkgo_omp -lginkgo_cuda -lginkgo_reference -lrt
> ```

it reports: 

/usr/bin/ld: ../../libmfem.a(mem_manager.o): undefined reference to symbol 'cudaStreamSynchronize@@libcudart.so.10.2'
//usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudart.so.10.2: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status

**Comment by tcojean** [comment]
I guess you need to add `-lcudart` to the previous command (possibly also `-L/path/to/libcudart.so` before). Depending on what launching the previous command with this addition tells you afterwards, `-lcublas` and `-lcusparse` are also possible.

Another solution is to try to extract the code from #586 (as a patch for example) and add it to the Ginkgo you are using...

**Comment by nghiakvnvsd** [comment]
> I guess you need to add `-lcudart` to the previous command (possibly also `-L/path/to/libcudart.so` before). Depending on what launching the previous command with this addition tells you afterwards, `-lcublas` and `-lcusparse` are also possible.
> 
> Another solution is to try to extract the code from #586 (as a patch for example) and add it to the Ginkgo you are using...

I did as you supported but It still report the error. I think I have to fix the code from #586 as you mentioned but when I saw all the code from #586, it's very different with my file (the number of line and some codes). I think maybe the reason is I use ginkgo 1.1.1?

**Comment by nghiakvnvsd** [comment]
I try to fix all codes from #586 (follow link: https://codecov.io/gh/ginkgo-project/ginkgo/pull/586/diff?src=pr&el=tree-more) but it still doesn't work.

**Comment by upsj** [comment]
@nghiakvnvsd we have not seen any activity on this issue for a while - is it still relevant?
